File: backend/app/schemas/entry.py
from datetime import datetime, timezone
from typing import Optional, List, Literal
from uuid import UUID
from pydantic import BaseModel, field_serializer


MoodType = Literal["great", "good", "okay", "bad", "terrible"]
JournalType = Literal["morning", "evening", "freeform"]


def serialize_datetime(dt: datetime) -> str:
    """Serialize datetime to ISO format with Z suffix for iOS compatibility."""
    if dt is None:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    utc_dt = dt.astimezone(timezone.utc)
    return utc_dt.strftime('%Y-%m-%dT%H:%M:%S') + 'Z'


class EntryCreate(BaseModel):
    title: Optional[str] = None
    content: str
    transcript: Optional[str] = None
    mood: Optional[MoodType] = None
    journal_type: Optional[JournalType] = None
    themes: Optional[List[str]] = None


class EntryUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    mood: Optional[MoodType] = None
    journal_type: Optional[JournalType] = None


class EntryResponse(BaseModel):
    id: UUID
    title: Optional[str]
    content: str
    transcript: Optional[str] = None
    mood: Optional[str]
    journal_type: Optional[str]
    themes: Optional[List[str]] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

    @field_serializer('created_at', 'updated_at')
    def serialize_dates(self, dt: datetime) -> str:
        return serialize_datetime(dt)


class SimilarEntryResponse(BaseModel):
    id: UUID
    title: Optional[str]
    content: str
    mood: Optional[str]
    created_at: datetime
    similarity: float

    @field_serializer('created_at')
    def serialize_created_at(self, dt: datetime) -> str:
        return serialize_datetime(dt)


class EntryListResponse(BaseModel):
    entries: List[EntryResponse]
    total: int
    page: int
    limit: int

File: ios/JournalBuddy/JournalBuddy/Models/Entry.swift
import Foundation

enum MoodType: String, Codable, CaseIterable {
    case great
    case good
    case okay
    case bad
    case terrible

    var emoji: String {
        switch self {
        case .great: return "😄"
        case .good: return "🙂"
        case .okay: return "😐"
        case .bad: return "😔"
        case .terrible: return "😢"
        }
    }

    var displayName: String {
        rawValue.capitalized
    }
}

enum JournalType: String, Codable, CaseIterable {
    case morning
    case evening
    case freeform

    var displayName: String {
        rawValue.capitalized
    }

    var icon: String {
        switch self {
        case .morning: return "sunrise.fill"
        case .evening: return "moon.stars.fill"
        case .freeform: return "pencil"
        }
    }
}

struct Entry: Codable, Identifiable {
    let id: UUID
    var title: String?
    var content: String
    var transcript: String?
    var mood: MoodType?
    var journalType: JournalType?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, title, content, transcript, mood
        case journalType = "journal_type"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct EntryCreate: Codable {
    var title: String?
    var content: String
    var transcript: String?
    var mood: MoodType?
    var journalType: JournalType?

    enum CodingKeys: String, CodingKey {
        case title, content, transcript, mood
        case journalType = "journal_type"
    }
}

struct EntryUpdate: Codable {
    var title: String?
    var content: String?
    var mood: MoodType?
    var journalType: JournalType?

    enum CodingKeys: String, CodingKey {
        case title, content, mood
        case journalType = "journal_type"
    }
}

struct EntryListResponse: Codable {
    let entries: [Entry]
    let total: Int
    let page: Int
    let limit: Int
}
